pkg/metrics/metrics.go

```diff
@@ -0,0 +1,163 @@
+	}, []string{"name", "namespace", "uid", "condition", "status"})
+
+	MachineSetCountDesc = prometheus.NewDesc("mcm_machineset_items_total", "Count of machinesets currently managed by the mcm.", nil, nil)
+	MachineSetCreated   = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_created",
+		Help: "Creation time of the Machinesets currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid"})
+
+	MachineSetInfo = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_info",
+		Help: "Information of the Machinesets currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid", "generation", "kind", "api_version",
+		"spec_machine_class_api_group", "spec_machine_class_kind", "spec_machine_class_name"})
+
+	MachineSetInfoSpecReplicas = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_info_spec_replicas",
+		Help: "enca countof the Machinesets currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid"})
+
+	MachineSetInfoSpecMinReadySeconds = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_info_spec_min_ready_seconds",
+		Help: "Information of the Machinesets currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid"})
+
+	MachineSetStatusCondition = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_status_condition",
+		Help: "Information of the mcm managed Machinesets' status conditions.",
+	}, []string{"name", "namespace", "uid", "condition", "status"})
+
+	MachineSetStatusFailedMachines = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_failed_machines",
+		Help: "Information of the mcm managed Machinesets' failed machines.",
+	}, []string{"name", "namespace", "uid", "failed_machine_name", "failed_machine_provider_id", "failed_machine_owner_ref",
+		"failed_machine_last_operation_state",
+		"failed_machine_last_operation_machine_operation_type"})
+
+	MachineSetStatus = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_set_status",
+		Help: "Information of the mcm managed Machinesets' status conditions.",
+	}, []string{"name", "namespace", "uid", "available_replicas",
+		"fully_labeled_replicas", "ready_replicas", "replicas"})
+
+	MachineDeploymentCountDesc = prometheus.NewDesc("mcm_machinedeployment_items_total", "Count of machinedeployments currently managed by the mcm.", nil, nil)
+	MachineDeploymentCreated   = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_deployment_created",
+		Help: "Creation time of the Machinedeployments currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid"})
+	MachineDeploymentInfo = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mcm_machine_deployment_info",
+		Help: "Information of the Machinedeployments currently managed by the mcm.",
+	}, []string{"name", "namespace", "uid", "generation", "kind", "api_version", "spec_replicas", "spec_strategy_type",
```

Comment:
Same for the `generation`. If needed, we should create a dedicated metric for that.

I would also create dedicated metrics for `spec_paused` e. g. `mcm_machine_deployment_paused{"name", namespace, "uid"} -> 0=paused, 1=unpaused`. It's the pretty similar for the other labels  `spec_strategy_rolling_update_max_surge`, `spec_strategy_rolling_update_max_unavailable`, `spec_min_ready_seconds`, `spec_min_ready_seconds`.

I think the `kind` label is not required, because the resource kind is already expressed through the metric name "mcm_machine_deployment_*". I'm also not sure about the `api_version` label.